examples/keras_io/tensorflow/structured_data/deep_neural_decision_forests.py

```diff
@@ -0,0 +1,459 @@
+            # (oov) token, we set mask_token to None and num_oov_indices to 0.
+            lookup = StringLookup(
+                vocabulary=vocabulary, mask_token=None, num_oov_indices=0
+            )
+            # Convert the string input values into integer indices.
+            value_index = lookup(inputs[feature_name])
+            embedding_dims = int(math.sqrt(lookup.vocabulary_size()))
+            # Create an embedding layer with the specified dimensions.
+            embedding = layers.Embedding(
+                input_dim=lookup.vocabulary_size(), output_dim=embedding_dims
+            )
+            # Convert the index values to embedding representations.
+            encoded_feature = embedding(value_index)
+        else:
+            # Use the numerical features as-is.
+            encoded_feature = inputs[feature_name]
+            if inputs[feature_name].shape[-1] is None:
+                encoded_feature = keras.ops.expand_dims(encoded_feature, -1)
+
+        encoded_features.append(encoded_feature)
+
+    encoded_features = layers.concatenate(encoded_features)
+    return encoded_features
+
+
+"""
+## Deep Neural Decision Tree
+
+A neural decision tree model has two sets of weights to learn. The first set is `pi`,
+which represents the probability distribution of the classes in the tree leaves.
+The second set is the weights of the routing layer `decision_fn`, which represents the probability
+of going to each leave. The forward pass of the model works as follows:
+
+1. The model expects input `features` as a single vector encoding all the features of an instance
+in the batch. This vector can be generated from a Convolution Neural Network (CNN) applied to images
+or dense transformations applied to structured data features.
+2. The model first applies a `used_features_mask` to randomly select a subset of input features to use.
+3. Then, the model computes the probabilities (`mu`) for the input instances to reach the tree leaves
+by iteratively performing a *stochastic* routing throughout the tree levels.
+4. Finally, the probabilities of reaching the leaves are combined by the class probabilities at the
+leaves to produce the final `outputs`.
+"""
+
+
+class NeuralDecisionTree(keras.Model):
+    def __init__(self, depth, num_features, used_features_rate, num_classes):
+        super().__init__()
+        self.depth = depth
+        self.num_leaves = 2**depth
+        self.num_classes = num_classes
+
+        # Create a mask for the randomly selected features.
+        num_used_features = int(num_features * used_features_rate)
+        one_hot = np.eye(num_features)
+        sampled_feature_indicies = np.random.choice(
+            np.arange(num_features), num_used_features, replace=False
+        )
+        self.used_features_mask = one_hot[sampled_feature_indicies]
+
+        # Initialize the weights of the classes in leaves.
+        self.pi = tf.Variable(
```

Comment:
I'm pretty sure this won't work -- the variable won't be tracked. Use `self.add_weight` instead.